### PR TITLE
Change how we assemble user names in ASC

### LINKED ISSUE #44: Credentials with dot ('.') in domain name are parsed improperly by get_enterprise_name
We use gssntlmssp for SMB server authentication with external user (both client and server are joined to AD, winbind is running).
When SMB user is using FQDN as domain name (with dots), gssntlmssp tries to authenticate wrong user and winbind fails the authentication.

Use-case:
Windows user is trying to map SMB drive manually via cmd.exe with custom '/user' parameter where domain is represented as FQDN:

**net use x: \\server_ip\share_name /user:my.domain.fqdn\username**

We have the following problematic flow:
**gssntlm_accept_sec_context -> gssntlm_import_name -> gssntlm_import_name_by_mech -> get_enterprise_name**

On the level of gssntlm_accept_sec_context we have useratdom='username@my.domain.fqdn' and it is parsed by get_enterprise_name as whole username because it contains dots in domain name. I.e. gssntlmssp parses it
as domain=NULL and username='username@my.domain.fqdn'.
Expected behaviour is to parse it as domain='my.domain.fqdn' and username='username'.

Note - using FQDN as domain name is required in cases when short domain name is not unique in AD forest, for example
we have 'sales.us.mycompany.com' and 'sales.europe.mycompany.com'; in this case short domain name 'sales' is not unique and cannot be used.

In the code I see following commit related to this bug (see below). When the change is reverted, the issue is fixed.
I think that we need to handle "email address as username" use-case in other more safe way (check only for \\@ in username, don't rely on dots in domain name) because dots in domain name is not "unlikely" but the required flow.

```
$ git show -p 12cfa317
commit 12cfa317b8cdee94d0c4a6daf30dec56cd8d8fe2
Author: Simo Sorce <simo@redhat.com>
Date:   Tue Jan 6 14:24:58 2015 -0500

    Names with a . in the domain are enteprise names
    
    This allows people to put in an email address as the source name and
    have i treated automatically as an enterprise name as well.
    
    Although technically NetBIOS names can have dots it is unlikely and the
    user@domain form is generally undestood to be used with UPNs and email
    like addresses which use the DNS Domain Name.
    
    The fallback case for NetBIOS domain names with a dot is to configure the
    client to use the DOMAIN\user name form instead.

diff --git a/src/gss_names.c b/src/gss_names.c
index e76a3e8..1a35ed2 100644
--- a/src/gss_names.c
+++ b/src/gss_names.c
@@ -96,11 +96,18 @@ static uint32_t get_enterprise_name(uint32_t *minor_status,
     buf[len] = '\0';
 
     e = strstr(buf, "\\@");
+    if (e) {
+        /* remove escape */
+        memmove(e, e + 1, len - (e - buf));
+    } else {
+        /* check if domain part contains dot */
+        e = strchr(buf, '@');
+        if (e) {
+            e = strchr(e, '.');
+        }
+    }
     if (!e) return GSSERRS(0, GSS_S_UNAVAILABLE);
 
-    /* remove escape */
-    memmove(e, e + 1, len - (e - buf));
-
     *username = strdup(buf);
     if (NULL == *username) {
         set_GSSERR(ENOMEM);
```

### PR DESCRIPTION
**Pull request by simo5**
Use a DOMAIN\Username format to avoid accidental detection of
a name that has a dotted domain as an enterprise name.

Fixes #44

### CODE REVIEW

**Review by frozencemetery — APPROVED** [review]

**Comment by simo5** [comment]
@frozencemetery @kvv81 I added another patch to handle this better on the get_enterprise_name() part as well.
Please (re)review.

**Review by frozencemetery — CHANGES_REQUESTED** [review]
When `get_enterprise_name()` is called, the (currently sole) callsite proceeds to re-check what format it's in.  I think a better way to structure this would be to combine these two into a single pass.

File: src/gss_names.c [hunk 1]
```diff
@@ -62,36 +62,137 @@ static uint32_t string_split(uint32_t *minor_status, char sep,
...
     if (len > MAX_NAME_LEN) {
         return GSSERRS(ERR_NAMETOOLONG, GSS_S_BAD_NAME);
     }
-    buf = alloca(len + 1);
 
+    /* let's check if there is an '@' sign at all */
+    at = memchr(str, '@', len);
+    if (!at) {
+        /* not an enterprise name for sure */
+        return GSSERRS(0, GSS_S_UNAVAILABLE);
```
[the remark this answers is not in the document]
Comment (simo5) [reply src/gss_names.c:120]: no, this is a valid name, just not an enteprise name, GSS_S_UNAVAILABLE, cause the code to continue with normal processing int he caller.

File: src/gss_names.c [hunk 2]
```diff
@@ -62,36 +62,137 @@ static uint32_t string_split(uint32_t *minor_status, char sep,
...
+    }
+
+    /* check if we have a '\' at all */
+    sep = memchr(str, '\\', len);
+    if (!sep) {
+        /* not an enterprise name for sure */
+        return GSSERRS(0, GSS_S_UNAVAILABLE);
```
[the remark this answers is not in the document]
Comment (simo5) [reply src/gss_names.c:127]: I guess, yeah.

File: src/gss_names.c [hunk 3]
```diff
@@ -62,36 +62,137 @@ static uint32_t string_split(uint32_t *minor_status, char sep,
...
+    }
+
+    /* copy buf to manipulate it */
+    buf = alloca(len + 1);
```
[the remark this answers is not in the document]
Comment (simo5) [reply src/gss_names.c:131]: Is there a problem with it ?
This is just a temporary buffer.

[the remark this answers is not in the document]
Comment (simo5) [reply src/gss_names.c:131]: I guess I can simply change it to a dynamically allocated array ... char buf[len+1]; ...

**Comment by simo5** [comment]
I wonder if I should get away with get_enterprise_name() and just make a parse_user_name() function that processes everything.

**Comment by simo5** [comment]
Ok I pushed a proposed fixup version where I get all username processing in a single parse_user_name() function.

Let me know if this looks better. The behavior is unchanged from the previous patch, it is just moving around stuff.

File: src/gss_names.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (frozencemetery) [reply src/gss_names.c]: Both of those are often prohibited in codebases due to potential for sudden very large stack allocations.  `alloca()` is often also prohibited because nothing prevents the value it returns from being returned somehow (and then becoming garbage).

Your usage is correct; I'm just surprised to see them, so I thought I'd check :)

**Comment by kvv81** [comment]
Oops, it seems like enterprise name use-case is not going well...
зображення

I see a problem with escaping in this flow:
1. gssntlm_accept_sec_context is called, security context has DOMAIN=(NULL) and user='username@gmail.com'
2. call to ntlm_decode_auth_msg is done, however this routine does not add escaping for (at) character, so on the output we have DOMAIN=(NULL) and user='username@gmail.com'.
3. During the call of gssntlm_import_name -> gssntlm_import_name_by_mech -> parse_user_name finds only 'at' character without 'sep' because escaping was not done; as a result, this use-case is handled as standard foo@BAR form.
4. winbind is called with wrong domain/username, NO_SUCH_USER error.
NTLM CRAP authentication for user [gmail.com]\[username] returned NT_STATUS_NO_SUCH_USER

**Comment by simo5** [comment]
Very good point, thanks for testing this is exactly the kind of validation I was hoping to see.

So we have two options here:
Always compose as if a domain was there, and treat \user@foo as valid with result:
  DOMAIN=NULL, USERNAME=user@foo
or always escape the username therefore allowing domain\user\@foo as valid with result:
  DOMAIN=domain, USERNAME=user@foo 

The first form has the advantage of avoiding escaping, and introduces a new way to pass in an enterprise name.
The second is more consistent with the current documentation, but requires internal escaping and a change in parse_user_name()

Finally we have a third option, reactitect ASC to be able to pass in a gssntlm_name instead and avoid translations, but this does not solve the case where a user need to handle this stuff from a client application.

I am leaning toward the first form because ti will make it easier to deal ith from client apps as well without escaping,
WDYT ?

**Comment by simo5** [comment]
I changed the code according to the preference I expressed in my previous comment.
@kvv81 it would be really nice if you could retest in the failing scenario to assure we properly handle it now.

**Comment by kvv81** [comment]
Well, I have tried to test all those corner-cases in-depth and got unexpected result...
Using special characters like (at) and (backslash) doesn't work well even in native Windows environment when using 'net use' command.

First of all, special characters are forbidden by ActiveDirectory itself, when trying to add such you will see this warning:
зображення

As the result, pre-win2k user logon name is fixed to NOT contain any (at) or (backslash). When 'DOMAIN\user' old form is used, special characters MUST be replaced with underscores.
New (unchanged) user logon name can be used with UPN form only (username@domain.com), BUT in this case special characters like (at) or (slash) are also forbidden for reason of collision with old format and (at) separator character.
I.e. the usage of new (post-win2k) user logon name is probably due unlimited length of user logon name (pre-win2k has a limit of 20 characters, new logon name list much longer - 256?) and maybe due other special characters. Although (at) and (backslash) characters are allowed in new logon name, you simply can't use it...

зображення

Bottom line - I do not see any benefit to support enterprise names like 'username@example.com' (username_example.com must be used instead). Could you please advise in what environment did you see this usage before? Probably my reproduction was not full enough (I used 'net use' and RemoteDesktop apps to check usernames with special chars - not handled there).

**Comment by kvv81** [comment]
As the test itself for your changes - I got a new problem with 'plain username' use-case in my scenario.
I.e. when I run "net use \\10.0.1.62\share /user:test_user", I get the error.

In tcpdump I see following things:
- There are TWO attempts of login, the 1st is done immediately (with default NULL domain and empty password?) and usually fails with LOGIN_FAILURE. From winbind side I see this line for 1st try:
**NTLM CRAP authentication for user [VAST-SMB-DEV]\[test_user] returned NT_STATUS_WRONG_PASSWORD**
- The 2nd try is done later (after correct password is entered) BUT domain name is "10.0.1.62" (set by client).
Winbind doesn't understand this trick and this try is failed too.
**NTLM CRAP authentication for user [10.0.1.62]\[test_user] returned NT_STATUS_NO_SUCH_USER**

зображення

I understand that here we have problem driven from client-side, but still we need to solve it somewhere.
The 1st try also fails on WindowsServer (due bad password, we can't do anything here).
But the 2nd try (with domain=server_ip) passes properly on windows. Probably we need to make some work-around here

**Comment by simo5** [comment]
The \test_user issue cropped up also in the test suite, and I have already rebased yesterday to fix it, can you check you got the latest release ?

**Comment by simo5** [comment]
In a domain named foo.com you can set a UPN like user@bar.com, in this case user@bar.com is an enterprise name, as it has nothing to do with the actual AD domain name. Enteprise names need special resolution in AD, which is why they are passed as USERNAME=user@bar.com DOMAIN=NULL

That's all there is, paring user@bar.com as USERNAME=user DOMAIN=bar.com would fail as there is no such bar.com AD (or NETBIOS) domain in the forest.

**Comment by simo5** [comment]
Btw the 10.0.1.62 "domain name" looks like a bug in whatever client you are using, I am not aware of that being a valid name at all, sounds like a resolution issue in some fallback.

**Comment by kvv81** [comment]
> The \test_user issue cropped up also in the test suite, and I have already rebased yesterday to fix it, can you check you got the latest release ?

Sorry, I don't understand what is the "latest release"...
I have got both commits from your enterprise_name_issues branch, and the latest commit from main is dated by September (also present).
What new code do you mean?

```
$ git branch
* enterprise_name_issues

$ git log

commit 2d5c7d1bc66e1d481b028b2c1dda2015b838ba14
Author: Simo Sorce <simo@redhat.com>
Date:   Mon Oct 12 16:32:54 2020 -0400

    Change how we assemble user names in ASC
...
commit 75ed138925c748622ef9cef51085d46747e4c092
Author: Simo Sorce <simo@redhat.com>
Date:   Tue Oct 13 14:36:55 2020 -0400

    Introduce parse_user_name() function
...
commit ecda1a5a82300b72af47260f1e90c181c04a2270
Author: Volodymyr Khomenko <volodymyr@vastdata.com>
Date:   Tue Sep 29 14:00:05 2020 +0300

    Fixed memory leaks found by valgrind
```

**Comment by simo5** [comment]
That's the right top commit, but now I am baffled.
In a previous push I had this issue that I was getting \username as well, so I rebased an push the current commit to properly address that.
Now the test suite shown no such  behavior, and it can't because the code properly parses a \username as USERNAME=username, DOMAIN=NULL and the \ is gone, unless somehow someone passes in \\username as a username ?

But that is not what you do in the Windows commandline, so I am baffled about how this can happen.

Would you be able to capture the NTLM negotiation before we hand data over to Winbind and see if the \ is already present in the username there? ATM I cannot explain how else a username can end up having a leading \ ...

**Comment by kvv81** [comment]
I'm afraid that UPN use-cases are still problematic with winbind.

This time I'm using this "happy path" matrix (it passes for "net use" command in case of Win-Server):
зображення

What I see that UPN scenarios are failed due splitting enterprise name on parts.
Note - for cases 2 vs 3 (and 4 vs 5) winbind receives exactly the same input data (domain and account),
also used password '12345' is the same, BUT we still have different auth result.
I think that password hash also depends on account name, so using improper account  name leads to wrong password result.

```
1. net use v: \\10.0.1.62\root 12345 /user:test_user

useratdom=\test_user
winbind_srv_auth: account_name=test_user, domain_name=VAST-SMB-DEV, workstation=CNTOS-VOLODYMR2
[1603112103] ALLOK: gssntlm_srv_auth() @ src/gss_auth.c:446 [0:0]

NTLM CRAP authentication for user [VAST-SMB-DEV]\[test_user] returned NT_STATUS_OK


2. net use v: \\10.0.1.62\root 12345 /user:vast-smb-dev.local\test_user

useratdom=vast-smb-dev.local\test_user
winbind_srv_auth: account_name=test_user, domain_name=vast-smb-dev.local, workstation=CNTOS-VOLODYMR2
[1603112266] ALLOK: gssntlm_srv_auth() @ src/gss_auth.c:446 [0:0]

NTLM CRAP authentication for user [vast-smb-dev.local]\[test_user] returned NT_STATUS_OK


3. net use v: \\10.0.1.62\root 12345 /user:test_user@vast-smb-dev.local

useratdom=\test_user@vast-smb-dev.local
winbind_srv_auth: account_name=test_user, domain_name=vast-smb-dev.local, workstation=CNTOS-VOLODYMR2
[1603112407] ERROR: gssntlm_srv_auth() @ src/gss_auth.c:420 [851968:13]

NTLM CRAP authentication for user [vast-smb-dev.local]\[test_user] returned NT_STATUS_WRONG_PASSWORD


4. net use v: \\10.0.1.62\root 12345 /user:vast-smb-dev\test_user

useratdom=vast-smb-dev\test_user
winbind_srv_auth: account_name=test_user, domain_name=vast-smb-dev, workstation=CNTOS-VOLODYMR2
[1603112492] ALLOK: gssntlm_srv_auth() @ src/gss_auth.c:446 [0:0]

NTLM CRAP authentication for user [vast-smb-dev]\[test_user] returned NT_STATUS_OK


5. net use v: \\10.0.1.62\root 12345 /user:test_user@vast-smb-dev

useratdom=\test_user@vast-smb-dev
winbind_srv_auth: account_name=test_user, domain_name=vast-smb-dev, workstation=CNTOS-VOLODYMR2
[1603112570] ERROR: gssntlm_srv_auth() @ src/gss_auth.c:420 [851968:13]

NTLM CRAP authentication for user [vast-smb-dev]\[test_user] returned NT_STATUS_WRONG_PASSWORD

```

**Comment by kvv81** [comment]
Please note that showing "\test_user" is a wireshark side-effect, it uses (backslash) as a separator to show domain + username in one field. This just means that domain=NULL and username=test_user.

I think that using domain=server_ip_address is some client-specific, and now this behaviour has done (so use-case with plain username as logon_name passes now).

**Comment by simo5** [comment]
So it looks like what you are saying is that Windows maintain an enterprise name as is, while Winbindd seem to somehow decide to "reparse" the username after we hand it in, and gets it wrong ?

**Comment by simo5** [comment]
Can you get debug level 3 logs from Winbind ?

there is a log line that should show username and domain as:
[<pid>] pam auth crap domain: <domain> user: <username>

I would like to see what you get there for the "enterprise name" cases.

**Comment by kvv81** [comment]
```
net use v: \\10.0.1.62\root 12345 /user:test_user@vast-smb-dev.local
```
log from gssntlmssp (+I have added few verbose print-outs there):
```
useratdom=\test_user@vast-smb-dev.local
[1603121394] ALLOK: parse_user_name() @ src/gss_names.c:206 [0:0]
winbind_srv_auth: account_name=test_user, domain_name=vast-smb-dev.local, workstation=CNTOS-VOLODYMR2
[1603121394] ERROR: gssntlm_srv_auth() @ src/gss_auth.c:420 [851968:13]
```
wibind logs:
```
[ 4795]: pam auth crap domain: [vast-smb-dev.local] user: test_user
NTLM CRAP authentication for user [vast-smb-dev.local]\[test_user] returned NT_STATUS_WRONG_PASSWORD
process_request_done: [4795:PAM_AUTH_CRAP]: NT_STATUS_WRONG_PASSWORD
```

**Comment by simo5** [comment]
> useratdom=\test_user@vast-smb-dev.local
> [1603121394] ALLOK: parse_user_name() @ src/gss_names.c:206 [0:0]
> winbind_srv_auth: account_name=test_user, domain_name=vast-smb-dev.local, workstation=CNTOS-VOLODYMR2
> [1603121394] ERROR: gssntlm_srv_auth() @ src/gss_auth.c:420 [851968:13]

This is surprising, so you are saying that gssntlmssp is failing to provide:
    USERNAME=test_user@vast-smb-dev.local
    DOMAIN=NULL
and is instead splitting and providing
    USERNAME=test_user
    DOMAIN=vast-smb-dev.local

**Comment by simo5** [comment]
I think the issue may be in winbindd_get_creds() which should be invoked by gss_acquired_cred_from() right before external_srv_auth(0 is called.

**Comment by simo5** [comment]
@kvv81 I added a tweak to how represent enterprise username internally, by changing from using a NULL domain to an empty "" domain.
Can you check if this makes things better ?
It should make us not try to set a domain in winbindd_get_creds() for example.

**Comment by kvv81** [comment]
No, the tweak has no effect.
winbindd_get_creds() is not a source of a problem because the name is split earlier by parse_user_name().
In this use-case we have both 'at' and 'sep' pointers set and sep < at so parse_user_name() goes deeply inside the complicated logic I don't fully understand. I'm afraid the code of parse_user_name has become too cumbersome and it's too hard to troubleshoot it.

I suggest to write a unit-test for all use-cases you wrote in the header of parse_user_name - it will be more easy to run a test with gdb than try to guess all issues from source-code...

Having those print-outs added:
```
--- a/src/gss_names.c
+++ b/src/gss_names.c
@@ -317,6 +317,11 @@ uint32_t gssntlm_import_name_by_mech(uint32_t *minor_status,
                                  input_name_buffer->length,
                                  &name->data.user.domain,
                                  &name->data.user.name);
+        if (retmaj == GSS_S_COMPLETE) {
+           gssntlm_debug_printf("parse_user_name result: domain=%s, name=%s\n",
+              (name->data.user.domain) ? name->data.user.domain : "<null>",
+               (name->data.user.name) ? name->data.user.name : "<null>");
+        }
     } else if (gss_oid_equal(input_name_type, GSS_C_NT_MACHINE_UID_NAME)) {
 
         name->type = GSSNTLM_NAME_USER;
diff --git a/src/gss_sec_ctx.c b/src/gss_sec_ctx.c
index 2464571..81ab805 100644
--- a/src/gss_sec_ctx.c
+++ b/src/gss_sec_ctx.c
@@ -878,6 +878,7 @@ uint32_t gssntlm_accept_sec_context(uint32_t *minor_status,
             memcpy(&useratdom[uadlen], usr_name, ulen);
             uadlen += ulen;
             useratdom[uadlen] = '\0';
+            gssntlm_debug_printf("useratdom=%s\n", useratdom);
 
             usrname.value = useratdom;
             usrname.length = uadlen;
```

Running the command:
net use v: \\10.0.1.62\root 12345 /user:test_user@vast-smb-dev.local

I get those logs:

```
useratdom=\test_user@vast-smb-dev.local
[1603127368] ALLOK: parse_user_name() @ src/gss_names.c:206 [0:0]
parse_user_name result: domain=vast-smb-dev.local, name=test_user
winbind_srv_auth: account_name=test_user, domain_name=vast-smb-dev.local, workstation=CNTOS-VOLODYMR2
[1603127368] ERROR: gssntlm_srv_auth() @ src/gss_auth.c:420 [851968:13]
```

**Comment by simo5** [comment]
Allright, I am guilty of not adding tests, that is bad, I just did not realize that parse_user_name() was failing at correctly parsing the name, my bad.
I'll add test cases before asking you for any more live tests.

**Comment by simo5** [comment]
Ok added unit tests, which revealed 2 issues in parse_user_name, one of which would cause the form \user@foo to not be properly recognized.
Now it should work as expected.

@kvv81  sorry for not adding this test earlier and  making you run in circles.

File: tests/ntlmssptest.c [hunk 4]
```diff
@@ -2702,6 +2702,92 @@ int test_ACQ_NO_NAME(void)
...
+                 strcmp(gss_username->data.user.domain,
+                        name_test[i].domain) != 0)) {
+                failed = true;
+            }
+        }
+
+        if (failed) {
+            fprintf(stderr, "gssntlm_import_name(%s) failed!\n",
+                    name_test[i].name);
+            fprintf(stderr, "Expected: [%s]\[%s]\n",
```
Comment (kvv81) [inline tests/ntlmssptest.c:2768]: Syntax highlighting says that "backslash should be doubled" if you want to print it, single backslash is just escaping for the next character :-)

**Review by kvv81 — CHANGES_REQUESTED** [review]

**Comment by kvv81** [comment]
Well, we are one step closer to the good result :-)
I see that parse_user_name now returns expected values. But as you mentioned before, winbindd_get_creds() is called after that and domain name is modified: NULL->default_winbind_domain. So probably winbind uses name as plain username and fails with NO_SUCH_USER.

Again, testing matrix is 5 use-cases that should all pass properly. 'Plain username' and 'old pre-win2k logon name' works properly, but 'UPN as login name' fails.
зображення

Problematic use-case:
net use v: \\10.0.1.62\root 12345 /user:test_user@vast-smb-dev.local

log from gssntlmssp (again, with some my instrumentation to track the issue):
```
useratdom=\test_user@vast-smb-dev.local
[1603178622] ALLOK: parse_user_name() @ src/gss_names.c:212 [0:0]
parse_user_name result: domain=<null>, name=test_user@vast-smb-dev.local
winbind_get_creds INPUT: name->data.user.domain=<null>, name->data.user.name=test_user@vast-smb-dev.local
winbind_get_creds OUTPUT: cred->cred.external.user.data.user.domain=VAST-SMB-DEV, cred->cred.external.user.data.user.name=test_user@vast-smb-dev.local
[1603178622] ALLOK: gssntlm_acquire_cred_from() @ src/gss_creds.c:500 [0:0]
winbind_srv_auth: account_name=test_user@vast-smb-dev.local, domain_name=VAST-SMB-DEV, workstation=CNTOS-VOLODYMR2
[1603178622] ERROR: gssntlm_srv_auth() @ src/gss_auth.c:420 [851968:13]
```

log from winbind:
```
process_request: Handling async request 3441:PAM_AUTH_CRAP
[ 3441]: pam auth crap domain: [VAST-SMB-DEV] user: test_user@vast-smb-dev

NTLM CRAP authentication for user [VAST-SMB-DEV]\[test_user@vast-smb-dev.local] returned NT_STATUS_NO_SUCH_USER
Finished processing child request 14
Writing 4008 bytes to parent
process_request_done: [3441:PAM_AUTH_CRAP]: NT_STATUS_NO_SUCH_USER
```

**Comment by kvv81** [comment]
> @kvv81 sorry for not adding this test earlier and making you run in circles.
Nevermind.

And we have finally came to the point when we rely on winbind's behaviour, this is tough topic for testing...
Thanks for helping us!

File: tests/ntlmssptest.c (hunk 4 above)
Comment (simo5) [reply tests/ntlmssptest.c:2768]: good catch! :)

**Comment by simo5** [comment]
I can try to change the code to not add the domain at all in the enterprise case, by passing in an empty domain.
But it is unclear if Winbindd will work properly.

Would you happen to have a capture of NTLM authentication done against a windows server with a UPN name ?
I want to check how windows deal with the user vs domain name in NTLM packets in this case.

**Comment by simo5** [comment]
I pushed a smal patch that should make winbindd_get_creds() not add a domain part.
Wonder if this is sufficient to get it working.

**Comment by kvv81** [comment]
> Would you happen to have a capture of NTLM authentication done against a windows server with a UPN name ?
> I want to check how windows deal with the user vs domain name in NTLM packets in this case.

Yes, I have reproduced all visible variants of logon name (down-level, UPN, plain) against WinServer before to verify that those cases are really supported. The result of captured traffic for SMB SESSION_SETUP is summarized in 'NTLM domain' field of the table:

зображення

For example, for logon_name='test_user@vast-smb-dev.local' you can see NTLM domain=NULL, meaning that domain name in SESSION_SETUP packet was set to NULL and the rest of information was passed as username (test_user@vast-smb-dev.local).

зображення

As a summary, I can guess following behaviour during SMB auth:
1. Client search for down-level format separator (backslash) in logon name. If found, the first part of logon name is encoded as domain, the rest comes to username (backslash is stripped).
2. If backslash is not found in logon name, SMB client DOES NOTHING with logon name (i.e. client doesn't parse domain from UPN because we may have some alternate UPN suffix that doesn't match to domain name). Thus UPN format is always come as is (by single piece) in username with domain=NULL.
3. (GUESS) SMB server (while handling NTLM authentication) is likely to pass NTLM domain and username as they are to Domain Controller (again, if UPN is inside NTLM username, it can be alternate UPN suffix a.k.a enterprise name that doesn't match to actual domain name, so SMB server can't predict correct domain name for UPNs).
4. Domain Controller uses NTLM Domain part as domain (if not NULL) and NTLM username as username. If domain is NULL, username can be either plain username or UPN (username@DOMAIN.fqdn or username@ALTERNATE.upn). Domain controller has all required information (short+fqdn for all DOMAINs, all the list of alternate UPN suffixes) to recognize username in UPN format and parsing is done properly here (on DC).

**Comment by kvv81** [comment]
And YES, this time all five cases works properly!

Now UPN works fine, plain username and down-level logon names are ok (as before).

For UPN use-case:
`net use v: \\10.0.1.62\root 12345 /user:test_user@vast-smb-dev.local`

I see logs from gssntlmssp:
```
useratdom=\test_user@vast-smb-dev.local
[1603276334] ALLOK: parse_user_name() @ src/gss_names.c:207 [0:0]
parse_user_name result: domain=, name=test_user@vast-smb-dev.local
winbind_get_creds INPUT: name->data.user.domain=, name->data.user.name=test_user@vast-smb-dev.local
winbind_get_creds OUTPUT: cred->cred.external.user.data.user.domain=, cred->cred.external.user.data.user.name=test_user@vast-smb-dev.local
[1603276334] ALLOK: gssntlm_acquire_cred_from() @ src/gss_creds.c:500 [0:0]
winbind_srv_auth: account_name=test_user@vast-smb-dev.local, domain_name=, workstation=CNTOS-VOLODYMR2
[1603276334] ALLOK: gssntlm_srv_auth() @ src/gss_auth.c:446 [0:0]
```

winbind log:
```
[ 9675]: pam auth crap domain: [] user: test_user@vast-smb-dev.local
NTLM CRAP authentication for user []\[test_user@vast-smb-dev.local] returned NT_STATUS_OK
```

**Comment by simo5** [comment]
Excellent, I'll integrate the wip commit then, and merge this PR.

**Comment by simo5** [comment]
@frozencemetery if you want to take a formal look, I think we are ready to merge.

**Review by frozencemetery — APPROVED** [review]
